Fb fm add sample to box automation

labkey-danield opened the pull request:
#### Rationale
Primarily had to update the EditableGrid to support FM AddSamples test. 

#### Related Pull Requests
* https://github.com/LabKey/inventory/pull/161

#### Changes
* Cleaned up some javadoc comments.
* Fixed check for read-only cells.
* Added EditableGrid.getCellValue to return the value of a cell.
* Added EditableGrid.getPickListForCell. Some cells have a pick-list this returns the displayed list.
* Clarified how the row number column is returned (the format is different in a location grid).
* Changed name of the function listOfnotPopulatedRows to listOfUnpopulatedRow in EditableGrid.
* Fixed a bug in assigning the metricUnit in SampleTypeDefinition.

labkey-danield committed:
Merge branch 'develop' into fb_fmAddSampleToBoxAutomation

labkey-danield committed:
Merge branch 'develop' into fb_fmAddSampleToBoxAutomation

labkey-danield committed:
Update editable grid to not ignore the row number column.
Update how editable grid identifies read-only cells.

labkey-danield committed:
Add convertLocationToCoordinates and convertCoordinatesToLocation in the BoxDisplay.
Add overloaded methods that take a string with a location value to BoxDisplay.
Renamed getTableCell to getCellWebElement method in BoxDisplay (updated files that used it).
Created FMPasteIntoGridPanel and added it to FMFreezerStorageViewPage.
Added getAlert and dismissAlert to FMFreezerStorageViewPage.
Added AddSamplesToStorageTest.
Updated some existing elementCache elements to be final.
Made SampleDetailPanel more robust by not having a hard fail if detail info is missing, now returns an empty list.
Added cane to single dimension storage location.
Using appropriate value for metric units when creating a SampleType by the API.
Update expected columns in SMSourcesCreateTest.

labkey-danield committed:
Change name of function to be grammatically correct.

labkey-danield committed:
Merge branch 'develop' into fb_fmAddSampleToBoxAutomation

labkey-danield committed:
Added a getCell to the EditableGrid.
Added/clarified some of the comments descriptions in the EditableGrid.
Moved some column names into enum in the the page.
Added a check to see 'Add Samples' is enabled.
Added a check to make sure that selectCellRange is complete before returning.
Add a function to return the available cells in a box.
Added more automated tests.

labkey-danield committed:
Cleaning up a logging message.
Added more tests, most notably a unit conversion test and error conditions test.
EditableGrid: cleaned up how a pick list is retrieved.
EditableGrid: Added function to get a filtered pick list.
EditableGrid: Do not dismiss the picklist by default.
EditableGrid: Add method to dismiss the pick list.
Added error banner component to the PasteIntoGridPanel.

labkey-danield committed:
Merge branch 'develop' into fb_fmAddSampleToBoxAutomation

labkey-danield committed:
Merge branch 'develop' into fb_fmAddSampleToBoxAutomation

labkey-danield committed:
Merge branch 'develop' into fb_fmAddSampleToBoxAutomation

labkey-chrisj commented on src/org/labkey/test/components/glassLibrary/grids/EditableGrid.java:
```diff
@@ -188,7 +203,7 @@ public int getRowCount()
     }
 
-    public List<Integer> listOfnotPopulatedRows()
+    public List<Integer> listOfUnPopulatedRows()
```
It may be preferable to call this getUnpopulatedRowIndices()

Also it will be helpful to document why this information is useful, and what exactly it means.  Perhaps the point to this code is to give callers an easy way to determine the index of the first empty row, in order to begin insert there?  If this is the case, we might be best-served to expose a getter for that index, and hide the internal workings of the component grid to prevent calling code from having to understand its inner workings

labkey-danield replied:
I've updated the function name to getEditableRowIndices, and the function below it to getReadonlyRowIndices and have added a javadoc above both.

I think that adding a getFirstEmptyRow would be useful.

The value in these functions, especially for Freezer Manager, is it returns a list of all the occupied spaces in a terminal location (readonly rows). This helps in validating that the grid reflects the same status as box or bag display.

labkey-chrisj commented on src/org/labkey/test/components/glassLibrary/grids/EditableGrid.java:
```diff
@@ -297,20 +340,129 @@ public void setCellValue(int row, String columnName, Object value)
+    }
+
+    /**
+     * For the given row get the values displayed in the picklist for the given column.
```
Are the pickList items the selectable options in a lookup list for the cell?  If so, it might be helpful to describe what a pickList is in those terms.

labkey-danield replied:
I changed pick-list to dropdown. Wikipedia shows the two terms as being [interchangeable](https://en.wikipedia.org/wiki/Drop-down_list#:~:text=A%20drop%2Ddown%20list%20(abbreviated,it%20displays%20a%20single%20value.).

labkey-chrisj commented on src/org/labkey/test/components/ui/EntityInsertPanel.java:
```diff
@@ -82,7 +82,7 @@ public EntityInsertPanel addRecords(List<Map<String, Object>> records)
-        List<Integer> rowIndices = elementCache().grid.listOfnotPopulatedRows();
+        List<Integer> rowIndices = elementCache().grid.listOfUnPopulatedRows();
 
         if(rowIndices.size() < records.size())
```
TBH I think the grid should handle questions like this one (do I have enough open rows to accept this data?) by itself.  This is outside the scope of the current PR probably, but it occurred to me just now.
Also FWIW, the grid usually (but sometimes doesn't) adds rows automatically if the user pastes a set of data that fits within the columnSet, but exceeds the current rowCount.

labkey-danield replied:
That is a bit beyond the scope of this pr, this was only a function name change.

Freezer Manager definitely falls into the category of being a fixed size grid and not growing to match the pasted data. The grid represents a fixed size storage container.

labkey-danield committed:
Merge branch 'develop' into fb_fmAddSampleToBoxAutomation

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/EditableGrid.java:
```diff
@@ -297,20 +340,129 @@ public void setCellValue(int row, String columnName, Object value)
+     * @param columnName The name of the column for the cell.
+     * @return A reference to this EditableGrid.
+     */
+    public EditableGrid dismissPickList(int row, String columnName)
+    {
+        WebElement gridCell = getCell(row, columnName);
+        Actions builder = new Actions(getDriver());
+        builder.sendKeys(gridCell, Keys.ESCAPE).build().perform();
+
+        return this;
+    }
```
Is there a way to make this work without targeting a specific cell? Just to simplify it, since there can only be one pick-list open (I presume). Maybe something like:
```suggestion
    public EditableGrid dismissPickList()
    {
        Actions builder = new Actions(getDriver());
        builder.sendKeys(getComponentElement(), Keys.ESCAPE).build().perform();

        return this;
    }
```

labkey-danield replied:
Changed.

labkey-danield committed:
Implementing code review feedback.
Changing the name of some of the new functions.
Adding better javadoc comments.
Removing un-needed parameters.
